File: Rust/decirational/src/lexer.rs
use crate::arithmetic::{is_cyclic_begin, is_cyclic_end, is_decimal_point, is_digit};
use crate::custom_integer::{CustomInteger, DError, DResult};
use crate::decimal_integer::strip_whitespace;
use crate::rational::parse_rational;
use crate::token::Token;
use std::marker::PhantomData;

/// Tokenizes an expression string: every structural character (one of
/// "()[]|+-*/%^") is always its own token, and every maximal run of number
/// characters (digits, '.', '{', '}') becomes one operand token. Carries no
/// state of its own - `T::from_i32`/`T::parse` (see `CustomInteger`) build
/// operands directly, so there's nothing to store per instance.
pub struct Lexer<T: CustomInteger>(PhantomData<T>);

fn is_number_char(c: u8) -> bool {
    is_digit(c) || is_decimal_point(c) || is_cyclic_begin(c) || is_cyclic_end(c)
}

impl<T: CustomInteger> Default for Lexer<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: CustomInteger> Lexer<T> {
    pub fn new() -> Self {
        Lexer(PhantomData)
    }

    fn parse_operand(&self, value: &str) -> DResult<Token<T>> {
        if let Ok(n) = value.parse::<i32>() {
            return Ok(Token::Integer(T::from_i32(n)));
        }
        if let Ok(large) = T::parse(value) {
            return Ok(Token::Integer(large));
        }
        match parse_rational(value) {
            Ok(r) => Ok(Token::Rational(r)),
            Err(_) => Err(DError::new(format!("{} is not a valid rational", value))),
        }
    }

    /// Tokenizes `expression`, returning an empty vector for blank input.
    pub fn get_tokens(&self, expression: &str) -> DResult<Vec<Token<T>>> {
        let expression = strip_whitespace(expression);
        if expression.is_empty() {
            return Ok(Vec::new());
        }
        let bytes = expression.as_bytes();
        let mut tokens = Vec::with_capacity(bytes.len());
        let mut i = 0;
        while i < bytes.len() {
            let c = bytes[i];
            // "//" (truncating integer division) needs one character of
            // lookahead, so it can't be a plain single-character match arm
            // like the rest of the structural tokens below.
            if c == b'/' {
                if i + 1 < bytes.len() && bytes[i + 1] == b'/' {
                    tokens.push(Token::IntegerDivide);
                    i += 2;
                } else {
                    tokens.push(Token::Divide);
                    i += 1;
                }
                continue;
            }
            let structural = match c {
                b'(' => Some(Token::LeftParen),
                b')' => Some(Token::RightParen),
                b'[' => Some(Token::LeftFloor),
                b']' => Some(Token::RightFloor),
                b'|' => Some(Token::AbsoluteBar),
                b'+' => Some(Token::Plus),
                b'-' => Some(Token::Minus),
                b'*' => Some(Token::Multiply),
                b'%' => Some(Token::Modulo),
                b'^' => Some(Token::Power),
                _ => None,
            };
            if let Some(tok) = structural {
                tokens.push(tok);
                i += 1;
                continue;
            }
            if is_number_char(c) {
                let start = i;
                while i < bytes.len() && is_number_char(bytes[i]) {
                    i += 1;
                }
                let value = &expression[start..i];
                tokens.push(self.parse_operand(value)?);
                continue;
            }
            return Err(DError::new(format!("illegal character {} in expression: {}", c as char, expression)));
        }
        Ok(tokens)
    }
}

File: Go/Decirational/lexer.go
package decirational

import (
	"fmt"
	"strconv"
)

// Lexer[T] tokenizes an expression string: every structural character (one
// of "()[]|+-*/%^") is always its own token, and every maximal run of number
// characters (digits, '.', '{', '}') becomes one operand token.
type Lexer[T CustomInteger[T]] struct {
	fromInt32 func(int32) T
	parseInt  func(string) (T, error)
}

// NewLexer builds a Lexer that constructs integer operands via fromInt32
// (for literals that fit a 32-bit int) and parseInt (for larger literals and
// as the fallback used to build the numerator/denominator of a rational
// literal).
func NewLexer[T CustomInteger[T]](fromInt32 func(int32) T, parseInt func(string) (T, error)) Lexer[T] {
	return Lexer[T]{fromInt32, parseInt}
}

func isNumberChar(c byte) bool {
	return isDigit(c) || isDecimalPoint(c) || isCyclicBeginChar(c) || isCyclicEndChar(c)
}

func (l Lexer[T]) parseOperand(value string) (Operand[T], error) {
	if n, err := strconv.ParseInt(value, 10, 32); err == nil {
		return Operand[T]{Type: OperandInteger, Large: l.fromInt32(int32(n))}, nil
	}
	if large, err := l.parseInt(value); err == nil {
		return Operand[T]{Type: OperandLargeInteger, Large: large}, nil
	}
	rat, err := ParseRational(value, l.parseInt)
	if err != nil {
		return Operand[T]{}, fmt.Errorf("%s is not a valid rational", value)
	}
	return Operand[T]{Type: OperandRational, Rat: rat}, nil
}

// GetTokens tokenizes expression, returning an empty slice for a nil/blank input.
func (l Lexer[T]) GetTokens(expression string) ([]Token, error) {
	expression = stripWhitespace(expression)
	if expression == "" {
		return []Token{}, nil
	}
	tokens := make([]Token, 0, len(expression))
	i := 0
	for i < len(expression) {
		c := expression[i]
		switch c {
		case '(':
			tokens = append(tokens, LeftParen)
			i++
		case ')':
			tokens = append(tokens, RightParen)
			i++
		case '[':
			tokens = append(tokens, LeftFloor)
			i++
		case ']':
			tokens = append(tokens, RightFloor)
			i++
		case '|':
			tokens = append(tokens, AbsoluteBar)
			i++
		case '+':
			tokens = append(tokens, OpPlus)
			i++
		case '-':
			tokens = append(tokens, OpMinus)
			i++
		case '*':
			tokens = append(tokens, OpMultiplication)
			i++
		case '/':
			if i+1 < len(expression) && expression[i+1] == '/' {
				tokens = append(tokens, OpIntegerDivision)
				i += 2
			} else {
				tokens = append(tokens, OpDivision)
				i++
			}
		case '%':
			tokens = append(tokens, OpModulo)
			i++
		case '^':
			tokens = append(tokens, OpPower)
			i++
		default:
			if !isNumberChar(c) {
				return nil, fmt.Errorf("illegal character %c in expression: %s", c, expression)
			}
			j := i
			for j < len(expression) && isNumberChar(expression[j]) {
				j++
			}
			operand, err := l.parseOperand(expression[i:j])
			if err != nil {
				return nil, err
			}
			tokens = append(tokens, operand)
			i = j
		}
	}
	return tokens, nil
}
